src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs

```diff
@@ -907,45 +908,36 @@ public static int Sum<T>(this ImmutableArray<T> items, Func<T, int> selector)
 
         internal static void CreateNameToMembersMap
             <TNamespaceOrTypeSymbol, TNamedTypeSymbol, TNamespaceSymbol>
-            (Dictionary<string, object> dictionary, Dictionary<String, ImmutableArray<TNamespaceOrTypeSymbol>> result)
+            (Dictionary<string, object> dictionary, Dictionary<string, ImmutableArray<TNamespaceOrTypeSymbol>> result)
             where TNamespaceOrTypeSymbol : class
             where TNamedTypeSymbol : class, TNamespaceOrTypeSymbol
             where TNamespaceSymbol : class, TNamespaceOrTypeSymbol
         {
             foreach (var kvp in dictionary)
-            {
-                object value = kvp.Value;
-                ImmutableArray<TNamespaceOrTypeSymbol> members;
+                result.Add(kvp.Key, createMembers(kvp.Value));
+
+            return;
 
-                var builder = value as ArrayBuilder<TNamespaceOrTypeSymbol>;
-                if (builder != null)
+            static ImmutableArray<TNamespaceOrTypeSymbol> createMembers(object value)
+            {
+                if (value is ArrayBuilder<TNamespaceOrTypeSymbol> builder)
                 {
                     Debug.Assert(builder.Count > 1);
-                    bool hasNamespaces = false;
-                    for (int i = 0; i < builder.Count; i++)
+                    foreach (var item in builder)
                     {
-                        if (builder[i] is TNamespaceSymbol)
-                        {
-                            hasNamespaces = true;
-                            break;
-                        }
+                        if (item is TNamespaceSymbol)
+                            return builder.ToImmutableAndFree();
                     }
 
-                    members = hasNamespaces
-                        ? builder.ToImmutable()
-                        : ImmutableArray<TNamespaceOrTypeSymbol>.CastUp(builder.ToDowncastedImmutable<TNamedTypeSymbol>());
-
-                    builder.Free();
+                    return ImmutableArray<TNamespaceOrTypeSymbol>.CastUp(builder.ToDowncastedImmutableAndFree<TNamedTypeSymbol>());
```

Comment:
if we get to end of the loop, we know we didn't have any namespaes, so no need to check any variablie and can just return teh downcasted values.